flink-autoscaler/src/main/java/org/apache/flink/autoscaler/tuning/MemoryTuning.java

```diff
@@ -80,57 +85,93 @@ public static Configuration tuneTaskManagerHeapMemory(
-        tuningConfig.set(TaskManagerOptions.TOTAL_PROCESS_MEMORY, totalMemory);
-        // Framework and Task heap memory configs add up together yield the max heap memory.
-        // To simplify the calculation, set the framework heap memory to zero.
-        tuningConfig.set(TaskManagerOptions.FRAMEWORK_HEAP_MEMORY, MemorySize.ZERO);
-        tuningConfig.set(TaskManagerOptions.TASK_HEAP_MEMORY, newHeapSize);
-
-        // All memory options which can be configured via fractions need to be set to their
-        // absolute values or, if there is no absolute setting, the fractions need to be
-        // re-calculated.
-        MemorySize managedMemory = memSpecs.getFlinkMemory().getManaged();
-        if (shouldTransferHeapToManagedMemory(config, heapDiffBytes)) {
-            // If RocksDB is configured, give back the heap memory as managed memory to RocksDB
-            MemorySize newManagedMemory =
-                    new MemorySize(managedMemory.getBytes() + Math.abs(heapDiffBytes));
-            LOG.info(
-                    "Increasing managed memory size from {} to {}",
-                    managedMemory,
-                    newManagedMemory);
-            tuningConfig.set(TaskManagerOptions.MANAGED_MEMORY_SIZE, newManagedMemory);
-        } else {
-            tuningConfig.set(TaskManagerOptions.MANAGED_MEMORY_SIZE, managedMemory);
-        }
-
-        tuningConfig.set(
+        var tuningConfig = new ConfigChanges();
+        // Adjust the total container memory
+        tuningConfig.addOverride(TaskManagerOptions.TOTAL_PROCESS_MEMORY, totalMemory);
+        // We do not set the framework/task heap memory because those are automatically derived from
+        // setting the other mandatory memory options for managed memory, network, metaspace and jvm
+        // overhead. However, we do precise accounting for heap memory above. In contrast to other
+        // memory pools, there are no fractional variants for heap memory. Setting the absolute heap
+        // memory options could cause invalid configuration states when users adapt the total amount
+        // of memory. We also need to take care to remove any user-provided overrides for those.
+        tuningConfig.addRemoval(TaskManagerOptions.TASK_HEAP_MEMORY);
+        // Set default to zero because we already account for heap via task heap.
+        tuningConfig.addOverride(TaskManagerOptions.FRAMEWORK_HEAP_MEMORY, MemorySize.ZERO);
+
+        MemorySize flinkMemorySize =
+                new MemorySize(
+                        memSpecs.getTotalFlinkMemorySize().getBytes() + flinkMemoryDiffBytes);
+
+        // All memory options which can be configured via fractions need to be re-calculated.
+        tuningConfig.addOverride(
+                TaskManagerOptions.MANAGED_MEMORY_FRACTION,
+                getFraction(newManagedSize, flinkMemorySize));
+        tuningConfig.addRemoval(TaskManagerOptions.MANAGED_MEMORY_SIZE);
```

Comment:
We remove them because if the user has them set, it will corrupt the memory configuration. If memory tuning is enabled, those configurations need to be removed. Once memory tuning is disabled again, the user-set values will be restored automatically. 

Generally, I've moved away from settings any absolute values. Fractions support the user updating the total memory without things breaking instantly.